File: frontend/lib/src/models/ChartDefinition.dart
import '../utils.dart';
import 'AnalysisDataPath.dart';

class ChartDefinition {
  final String id;
  final String title;
  final List<AnalysisDataPath> analysisDataPaths;
  final ChartType chartType;
  final bool isXDateTime;

  ChartDefinition(this.title, this.analysisDataPaths, this.chartType, this.isXDateTime, { this.id = "" });

  ChartDefinition.fromMap(Map data) :
        id = data["id"],
        title = data["title"],
        analysisDataPaths = (data["analysis_data_paths"] as List<dynamic>).map((x) => AnalysisDataPath.fromMap(x)).toList(),
        chartType = getEnumByString(ChartType.values, data["chart_type"], ChartType.Line),
        isXDateTime = data["is_x_datetime"];
}

enum ChartType {
  Line,
  Pie,
  Bar,
  Scatter,

  PostsFrequency,
  LanguageFrequency,
  AuthorFrequency
}

File: backend/Domain/Models/Storage/Response/ChartDefinition.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;

namespace Socneto.Domain.Models.Storage.Response
{
    public class ChartDefinition
    {
        [JsonProperty("id")]
        public Guid Identifier { get; set; }
        
        [JsonProperty("title")]
        public string Title { get; set; }
        
        [JsonProperty("analysis_data_paths")]
        public List<AnalysisDataPath> AnalysisDataPaths { get; set; }
        
        [JsonConverter(typeof(StringEnumConverter))]
        [JsonProperty("chart_type")]
        public ChartType ChartType { get; set; }
        
        [JsonProperty("is_x_post_datetime")]
        public bool IsXPostDatetime { get; set; }
    }

    public class AnalysisDataPath
    {
        [JsonProperty("property")]
        public string Property { get; set; }
        
        [JsonProperty("analyser_id")]
        public string AnalyserComponentId { get; set; }
    }

    public enum ChartType
    {
        Line,
        Pie,
        Bar,
        Table,
        WordCloud,
        Scatter,
        
        PostsFrequency,
        LanguageFrequency,
        AuthorFrequency
    }
}
